pkg/kademlia/kademlia.go

```diff
@@ -238,31 +240,38 @@ func (k *Kad) Start(ctx context.Context) error {
 }
 
 func (k *Kad) connectBootnodes(ctx context.Context) {
-	var count int
+	var attempts, connected int
+	var totalAttempts = maxBootnodeAttempts * len(k.bootnodes)
+
 	for _, addr := range k.bootnodes {
-		if count >= 3 {
+		if attempts >= totalAttempts || connected >= 3 {
 			return
 		}
 
 		if _, err := p2p.Discover(ctx, addr, func(addr ma.Multiaddr) (stop bool, err error) {
 			k.logger.Tracef("connecting to bootnode %s", addr)
+			if attempts >= maxBootnodeAttempts {
```

Comment:
In the case where we have a problem with `bootnodes`, and some "fresh" node cannot connect to any, it will still end up in a loop, trying to connect to some boot node. This is done by the `kademlia.manage` loop, which is fine, because if there the node is not connected to any peer, it should try to connect to some.

Reply:
yes, that is correct, and the connection will be tried periodically. i don't see a reason why we should keep on trying over and over again (without any delay either). also, the other problem is that even if some other node dialed in to the node which is trying to connect to the bootnodes (and for some reason cannot), will never go out of the bootnodes loop, although some other peers may already be connected.